homeassistant/components/generic_thermostat/climate.py

```diff
@@ -132,336 +333,529 @@ class GenericThermostat(ClimateDevice, RestoreEntity):
-        return [PRESET_NONE, PRESET_AWAY] if self._away_temp else PRESET_NONE
+            if old_preset_mode is not None and old_preset_mode in self.preset_modes:
+                self._preset_mode = old_preset_mode
+
+            if old_hvac_mode is not None and old_hvac_mode in self.hvac_modes:
+                self._hvac_mode = old_hvac_mode
+
+                # Restore the target temperature
+                if old_hvac_mode == HVAC_MODE_COOL:
+                    min_temp = self._cool_conf[CONF_HVAC_MODE_MIN_TEMP]
+                    max_temp = self._cool_conf[CONF_HVAC_MODE_MAX_TEMP]
+                    if (
+                        old_temperature is not None
+                        and min_temp <= old_temperature <= max_temp
+                    ):
+                        self._target_temp_cool = old_temperature
+                elif old_hvac_mode == HVAC_MODE_HEAT:
+                    min_temp = self._heat_conf[CONF_HVAC_MODE_MIN_TEMP]
+                    max_temp = self._heat_conf[CONF_HVAC_MODE_MAX_TEMP]
+                    if (
+                        old_temperature is not None
+                        and min_temp <= old_temperature <= max_temp
+                    ):
+                        self._target_temp_heat = old_temperature
+
+        await self._async_operate()
+        await self.async_update_ha_state()
 
     async def async_set_hvac_mode(self, hvac_mode):
         """Set hvac mode."""
-        if hvac_mode == HVAC_MODE_HEAT:
-            self._hvac_mode = HVAC_MODE_HEAT
-            await self._async_control_heating(force=True)
-        elif hvac_mode == HVAC_MODE_COOL:
-            self._hvac_mode = HVAC_MODE_COOL
-            await self._async_control_heating(force=True)
-        elif hvac_mode == HVAC_MODE_OFF:
-            self._hvac_mode = HVAC_MODE_OFF
-            if self._is_device_active:
-                await self._async_heater_turn_off()
-        else:
+        # No changes have been made
+        if self._hvac_mode == hvac_mode:
+            return
+        if hvac_mode not in self.hvac_modes:
             _LOGGER.error("Unrecognized hvac mode: %s", hvac_mode)
             return
+        _LOGGER.debug("HVAC mode changed to %s", hvac_mode)
+        self._hvac_mode = hvac_mode
+        await self._async_operate()
+
         # Ensure we update the current operation after changing the mode
         self.schedule_update_ha_state()
 
     async def async_set_temperature(self, **kwargs):
         """Set new target temperature."""
         temperature = kwargs.get(ATTR_TEMPERATURE)
-        if temperature is None:
+        hvac_mode = kwargs.get(ATTR_HVAC_MODE)
+
+        if hvac_mode is None:
+            hvac_mode = self._hvac_mode
+        elif hvac_mode not in self.hvac_modes:
+            _LOGGER.warning(
+                "Try to update temperature to %s for mode %s but this mode is not enabled. Skipping.",
```

Comment:
We don't end logging messages with period.
```suggestion
                "Ignoring update temperature to %s for mode %s as this mode is not enabled,
```